Fix vpMomentAlpha computation

xoox opened the pull request:
Since `atan()` returns angle in [-pi/2, pi/2], the current computed
angle is actually within [-pi/4, pi/4]. And the result is either the
major axis or the minor axis which is not specified.

This patch will always give out the angle of the major axis. A span of
[-pi/2, pi/2] is achieved by considering both sine and cosine of
2*alpha.

Please see http://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/OWENS/LECT2/node3.html for details.

`copysign()` of C++11 is used, and both clang and gcc don't complain
about it when `USE_CXX11` is switched off. If required by older
compilers, I'll add fixes for C++ standards before C++11.
![sine-cosine](https://user-images.githubusercontent.com/1014277/49499871-34dfa180-f8aa-11e8-993a-1c475f6fdd05.png)

fspindle commented:
Thanks for your PR.
We will need time to investigate...

xoox commented:
Test code.

```cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <visp3/core/vpMomentAlpha.h>
#include <visp3/core/vpMomentCentered.h>
#include <visp3/core/vpMomentDatabase.h>
#include <visp3/core/vpMomentGravityCenter.h>
#include <visp3/core/vpMomentObject.h>
#include <visp3/core/vpPoint.h>
#include <visp3/io/vpImageIo.h>
#include <visp3/io/vpParseArgv.h>

int main(int argc, const char **argv)
{
  char *ImgName;
  // Parse the command line to set the variables
  vpParseArgv::vpArgvInfo argTable[] = {
      {"-q", vpParseArgv::ARGV_STRING, (char *)NULL, (char *)&ImgName, "image file name."},
      {"-h", vpParseArgv::ARGV_HELP, (char *)NULL, (char *)NULL, "Print the help."},
      {(char *)NULL, vpParseArgv::ARGV_END, (char *)NULL, (char *)NULL, (char *)NULL}};
  // Read the command line options
  if (vpParseArgv::parse(&argc, argv, argTable, vpParseArgv::ARGV_NO_LEFTOVERS | vpParseArgv::ARGV_NO_ABBREV |
                                                    vpParseArgv::ARGV_NO_DEFAULTS)) {
    return (-1);
  }

  vpImage<unsigned char> I;
  vpImageIo::read(I, ImgName);

  vpMomentObject obj(3);
  obj.setType(vpMomentObject::DENSE_FULL_OBJECT);
  obj.fromImage(I, 127, vpCameraParameters()); // Init the dense object with the image
  vpMomentDatabase db;                         // database
  vpMomentGravityCenter g;                     // declaration of gravity center
  vpMomentCentered mc;                         // centered moments
  vpMomentAlpha alphaRef;                      // declare alpha as reference
  g.linkTo(db);                                // add gravity center to database
  mc.linkTo(db);                               // add centered moments
  alphaRef.linkTo(db);                         // add alpha depending on centered moments
  db.updateAll(obj);                           // All of the moments must be updated, not just alpha
  g.compute();                                 // compute the moment
  mc.compute();                                // compute centered moments AFTER gravity center
  alphaRef.compute();                          // compute alpha AFTER centered moments.
  std::cout << "(Ref) orientation = " << vpMath::deg(alphaRef.get()) << std::endl;
  return 0;
}
```

With the above code, test was conducted with the following images. Wee can see the current implementation didn't give out the expected orientation angle. In some cases it computed the orientation of minor-axis instead of major-axis.

image                                                                                                              | expected | current   | this PR
-------------------------------------------------------------------------------------------------------------------|----------|-----------|----------
![arrow+000](https://cloud.githubusercontent.com/assets/1014277/21538116/8dc2af02-cdd4-11e6-8674-920924e7c2c0.png) | 0        | -0.128108 | -0.128108
![arrow+045](https://cloud.githubusercontent.com/assets/1014277/21538120/94eab55e-cdd4-11e6-9784-9b0b51d18476.png) | 45       | 44.8881   | 44.8881
![arrow+090](https://cloud.githubusercontent.com/assets/1014277/21538128/a2f58be2-cdd4-11e6-83ab-98b1f94aed03.png) | ±90      | **_-0.128108_** | 89.8719
![arrow+135](https://cloud.githubusercontent.com/assets/1014277/21538131/ab84446a-cdd4-11e6-8ba8-d97159eaf35f.png) | -45      | **_44.8881_**   | -45.1119
![arrow+180](https://cloud.githubusercontent.com/assets/1014277/21538132/b2e3ccbc-cdd4-11e6-900a-6330a02ad9d6.png) | 0        | -0.128108 | -0.128108
![arrow-135](https://cloud.githubusercontent.com/assets/1014277/21538137/c6c68526-cdd4-11e6-8f25-053c170fcd36.png) | 45       | 44.8881   | 44.8881
![arrow-090](https://cloud.githubusercontent.com/assets/1014277/21538143/d36ff438-cdd4-11e6-98f7-25f5eb254f33.png) | ±90      | **_-0.128108_** | 89.8719
![arrow-045](https://cloud.githubusercontent.com/assets/1014277/21538145/de88043c-cdd4-11e6-901c-e99bc1418590.png) | -45      | **_44.8881_**   | -45.1119

fspindle commented:
These are exactly the same results as the one we had originally with
```
double alpha = 0.5 * atan2(2.0 * momentCentered.get(1, 1),  (momentCentered.get(2, 0) - momentCentered.get(0, 2)));
```

fspindle commented:
Just to remmember, atan2() usage was removed in commit commit 64937a5e in order to fix issue #117

xoox commented:
Thanks for the comments. I forgot the existence of `atan2()`. ;-)

I think we should revert back to the usage of `atan2()`. The issue of
out of range [-Pi, Pi] in relative mode (#117) was fixed in 64937a5. But
64937a5 also introduced a bug in reference mode computation. The new
added commit of this PR reverts back to atan2() usage in reference mode
computation of vpMomentAlpha. Issue #117 was already fixed by the
updated `if (signChange)` block.

With the following test code, I got the expected results for the above 8
images. The first column is reference mode orientation. The second
column is relative mode orientation which is exactly the arrow's
direction.

Results.
```
orientation = (Ref) -0.128108   (Rel) -0.128108
orientation = (Ref) 44.8881     (Rel) 44.8881
orientation = (Ref) 89.8719     (Rel) 89.8719
orientation = (Ref) -45.1119    (Rel) 134.888
orientation = (Ref) -0.128108   (Rel) 179.872
orientation = (Ref) 44.8881     (Rel) -135.112
orientation = (Ref) 89.8719     (Rel) -90.1281
orientation = (Ref) -45.1119    (Rel) -45.1119
```

Test code.
```cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <visp3/core/vpMomentAlpha.h>
#include <visp3/core/vpMomentCentered.h>
#include <visp3/core/vpMomentDatabase.h>
#include <visp3/core/vpMomentGravityCenter.h>
#include <visp3/core/vpMomentObject.h>
#include <visp3/core/vpPoint.h>
#include <visp3/io/vpImageIo.h>
#include <visp3/io/vpParseArgv.h>

// generic function for printing
void print(double i) { std::cout << i << "\t"; }

int main(int argc, const char **argv)
{
  char *ImgName;
  // Parse the command line to set the variables
  vpParseArgv::vpArgvInfo argTable[] = {
      {"-q", vpParseArgv::ARGV_STRING, (char *)NULL, (char *)&ImgName, "image file name."},
      {"-h", vpParseArgv::ARGV_HELP, (char *)NULL, (char *)NULL, "Print the help."},
      {(char *)NULL, vpParseArgv::ARGV_END, (char *)NULL, (char *)NULL, (char *)NULL}};
  // Read the command line options
  if (vpParseArgv::parse(&argc, argv, argTable, vpParseArgv::ARGV_NO_LEFTOVERS | vpParseArgv::ARGV_NO_ABBREV |
                                                    vpParseArgv::ARGV_NO_DEFAULTS)) {
    return (-1);
  }

  vpImage<unsigned char> I;
  vpImageIo::read(I, ImgName);

  vpMomentObject obj(3); // Must be of order 3
  // because we will need the 3rd order
  // centered moments
  obj.setType(vpMomentObject::DENSE_FULL_OBJECT);
  obj.fromImage(I, 127, vpCameraParameters()); // Init the dense object with the image
  vpMomentDatabase db;                         // database
  vpMomentGravityCenter g;                     // declaration of gravity center
  vpMomentCentered mc;                         //  centered moments
  vpMomentAlpha alphaRef;                      // declare alpha as reference
  std::vector<double> ref = {18.15916330118853, 0.5262041890075579, -6.17050894899765, 3.339821107883651};
  vpMomentAlpha alpha(ref, M_PI); // alpha relative
  g.linkTo(db);                   // add gravity center to database
  mc.linkTo(db);                  // add centered moments
  alphaRef.linkTo(db);            // add alpha depending on centered moments
  alpha.linkTo(db);               // add alpha depending on centered moments
  db.updateAll(obj);              // All of the moments must be updated, not just alpha
  g.compute();                    // compute the moment
  mc.compute();                   // compute centered moments AFTER gravity center
  alphaRef.compute();             // compute alpha AFTER centered moments.
  alpha.compute();                // compute alpha AFTER centered moments.
  std::cout << "orientation = "
            << "(Ref) " << vpMath::deg(alphaRef.get()) << "\t"
            << "(Rel) " << vpMath::deg(alpha.get()) << std::endl;
  return 0;
}
```

xoox commented:
Because `arctan` is in the interval of (-pi/2, pi/2], The usage of
`arctan` in the documentation formula is misleading. We can use directly
`atan2` which is in the interval of (-pi, pi]. There is also a page
about `atan2(y, x)` on Wikipedia https://en.wikipedia.org/wiki/Atan2

fspindle commented:
I agree, we should revert to `atan2()` usage.

I'm now just confused by the lines
```
std::vector<double> ref = {18.15916330118853, 0.5262041890075579, -6.17050894899765, 3.339821107883651};
vpMomentAlpha alpha(ref, M_PI); // alpha relative
```
How do you get the values of `ref` for the arrow angle of pi ?
When I print the values of `mu` for the left arrow image 
```
    mc.compute(); 
    alpha_ref.compute();
    alpha.compute(); 
    std::cout << "mu30=" << mc.get(3,0) << " mu21=" << mc.get(2,1) << " mu12=" << mc.get(1,2) << " mu03=" << mc.get(0,3) << std::endl;
```
I got:
```
mu30=1.57498e-05 mu21=8.49976e-10 mu12=-3.12456e-07 mu03=3.87505e-09
```
which are not the values set in `ref`

xoox commented:
`ref` was not computed based on the provided images. The images seem not appropriate to be used as references because their third order centered moments are too small so that `symmetric = true` will be set in the constructor.

The code was largely copied from #117 where the reference alpha was computed from some random points with code snippet like the following.
```cpp
    vpUniRand r(time(NULL));
    std::vector<vpPoint> FOVpoints;
    vpPoint p;
    int numPoints = 16 * r() + 16;

    for (int i = 0; i < numPoints; i++) {
        p.set_x(7. * r());
        p.set_y(3. * r());
        FOVpoints.push_back(p);
    }

    mObj.setType(vpMomentObject::vpObjectType::DISCRETE);
    mObj.fromVector(FOVpoints);
    mDb.updateAll(mObj);
    mGCenter.compute();
    mCentered.compute();
    mAlpha.compute();
    ref.clear();
    ref.push_back(mCentered.get(3, 0));
    ref.push_back(mCentered.get(2, 1));
    ref.push_back(mCentered.get(1, 2));
    ref.push_back(mCentered.get(0, 3));
```

Code in #117 uses the computed reference alpha (mAlpha.get(), which is -0.03803650948393809).
But in the code of my last comment I set `alpha_ref` to a manually chosen value. The relative angles are measured with regard to a user specified reference angle, which makes their values comparable with reference mode values.

In theory `alpha_ref ` in `vpMomentAlpha(const std::vector< double > &ref, double alphaRef)` seems not necessarily fixed to `ref` moments. Is my understanding right?

fspindle commented:
In `vpMomentAlpha(const std::vector< double > &ref, double alphaRef)` the  values of `ref` and `alphaRef` are linked. I just introduce new fixes in PR #482. Now results are correct even with small values of the 3rd order `mu` moments. There is also a new test that shows how to use `vpMomentAlpha` in relative mode.

From now see PR #482.
